gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -30,6 +33,90 @@ make_string (Location locus, std::string value)
     new AST::LiteralExpr (value, AST::Literal::STRING,
 			  PrimitiveCoreType::CORETYPE_STR, {}, locus));
 }
+
+/* Parse a single string literal from the given delimited token tree,
+   and return the LiteralExpr for it. Allow for an optional trailing comma,
+   but otherwise enforce that these are the only tokens.  */
+
+std::unique_ptr<AST::LiteralExpr>
+parse_single_string_literal (AST::DelimTokenTree &invoc_token_tree)
+{
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (std::move (lex));
+
+  auto last_token_id = TokenId::RIGHT_CURLY;
+  switch (invoc_token_tree.get_delim_type ())
+    {
+    case AST::DelimType::PARENS:
+      last_token_id = TokenId::RIGHT_PAREN;
+      rust_assert (parser.skip_token (LEFT_PAREN));
+      break;
+
+    case AST::DelimType::CURLY:
+      rust_assert (parser.skip_token (LEFT_CURLY));
+      break;
+
+    case AST::DelimType::SQUARE:
+      last_token_id = TokenId::RIGHT_SQUARE;
+      rust_assert (parser.skip_token (LEFT_SQUARE));
+      break;
+    }
+
+  rust_assert (parser.peek_current_token ()->get_id () == STRING_LITERAL);
+  std::unique_ptr<AST::LiteralExpr> lit_expr = parser.parse_literal_expr ();
+
+  parser.maybe_skip_token (COMMA);
+  rust_assert (parser.skip_token (last_token_id));
```

Comment:
If the user passes an incorrect amount of arguments, this will ICE while we could have a nicer formatting imo

Reply:
Yes, this was silly. We shouldn't ICE. I cleaned up the parsing and error handling here so now we get clear errors instead.